File: 19/16a.py
from itertools import cycle, islice

import numpy as np

v = np.array(list(map(int, open('16.txt').read())))
p = np.array(
    [
        list(islice(cycle(np.repeat([0, 1, 0, -1], i + 1)), 1, len(v) + 1))
        for i in range(len(v))
    ]
)
phase = lambda n: v if n == 0 else np.mod(np.abs(np.matmul(p, phase(n - 1))), 10)
print(''.join(map(str, phase(100)[:8])))

File: 19/16a.clj
(ns aoc (:require [clojure.core.matrix :refer [mmul]]))

(defn repeatelems [n seq]
  (mapcat (partial repeat n) seq))

(defn vabs [v]
  (map #(Math/abs %) v))

(defn vmod [m n]
  (map #(mod % n) m))

(def v (map (comp read-string str) (slurp "16.txt")))

(def p
  (for [i (range (count v))]
    (take (count v) (drop 1 (cycle (repeatelems (inc i) [0 1 0 -1]))))))

(defn phase [n]
  (if (= n 0)
    v
    (vmod (vabs (mmul p (phase (- n 1)))) 10)))

(apply str (take 8 (phase 100)))
